sim/firesim-lib/src/main/cc/bridges/tracerv.cc

```diff
@@ -232,64 +228,100 @@ size_t tracerv_t::process_tokens(int num_beats, int minimum_batch_beats) {
   // does not create a tracefile when trace_enable is disabled, but the
   // TracerV bridge still exists, and no tracefile is created by default.
   if (this->tracefile) {
-    if (this->human_readable || this->test_output) {
-      for (int i = 0; i < (bytes_received / sizeof(uint64_t)); i += 8) {
-        if (this->test_output) {
-          fprintf(this->tracefile, "%016lx", OUTBUF[i + 7]);
-          fprintf(this->tracefile, "%016lx", OUTBUF[i + 6]);
-          fprintf(this->tracefile, "%016lx", OUTBUF[i + 5]);
-          fprintf(this->tracefile, "%016lx", OUTBUF[i + 4]);
-          fprintf(this->tracefile, "%016lx", OUTBUF[i + 3]);
-          fprintf(this->tracefile, "%016lx", OUTBUF[i + 2]);
-          fprintf(this->tracefile, "%016lx", OUTBUF[i + 1]);
-          fprintf(this->tracefile, "%016lx\n", OUTBUF[i + 0]);
-          // At least one valid instruction
-        } else {
-          for (int q = 0; q < max_core_ipc; q++) {
-            if (OUTBUF[i + q + 1] & valid_mask) {
-              fprintf(this->tracefile,
-                      "Cycle: %016" PRId64 " I%d: %016" PRIx64 "\n",
-                      OUTBUF[i + 0],
-                      q,
-                      OUTBUF[i + q + 1] & (~valid_mask));
-            } else {
-              break;
-            }
+    std::function<void(uint64_t, uint64_t)> addInstruction = NULL;
+
+    if (fireperf) {
+      addInstruction = std::bind(&TraceTracker::addInstruction,
+                                 this->trace_tracker,
+                                 std::placeholders::_1,
+                                 std::placeholders::_2);
+    }
+    serialize(OUTBUF,
+              bytes_received,
+              tracefile,
+              addInstruction,
+              max_core_ipc,
+              human_readable,
+              test_output,
+              fireperf);
+  }
+  return bytes_received;
+}
+
+void tracerv_t::serialize(
+    const uint64_t *const OUTBUF,
+    const size_t bytes_received,
+    FILE *tracefile,
+    std::function<void(uint64_t, uint64_t)> addInstruction,
+    const int max_core_ipc,
+    const bool human_readable,
+    const bool test_output,
+    const bool fireperf) {
```

Comment:
Isn't fireperf tied to `addInstruction` somehow? Could we get rid of this boolean flag and rely on `addInstruction` being set or not? Also, would it be possible to pull the `addInstruction` logic completely out? It is not required for the serialization of data coming from the test.

Reply:
refactoring the body of serialize into using modern std::vector is outside the scope of this PR